core/src/main/java/io/questdb/cairo/CairoMetadata.java

```diff
@@ -0,0 +1,328 @@
+     */
+    public void hydrateTable(@NotNull TableToken token, @NotNull CairoConfiguration configuration, @NotNull Path path, @NotNull ColumnVersionReader columnVersionReader, boolean blindUpsert, boolean infoLog) {
+        if (infoLog) {
+            LOG.info().$("hydrating metadata [table=").$(token.getTableName()).I$();
+        }
+
+        // set up dir path
+        path.of(configuration.getRoot())
+                .concat(token.getDirName());
+
+        boolean isSoftLink = Files.isSoftLink(path.$());
+
+        // set up table path
+        path.concat(TableUtils.META_FILE_NAME)
+                .trimTo(path.size());
+
+        MemoryCMR metaMem = Vm.getCMRInstance();
+
+        try {
+            // open metadata
+            metaMem.smallFile(configuration.getFilesFacade(), path.$(), MemoryTag.NATIVE_METADATA_READER);
+            TableUtils.validateMeta(metaMem, null, ColumnType.VERSION);
+
+            // create table to work with
+            CairoTable table = new CairoTable(token);
+            table.setMetadataVersion(Long.MIN_VALUE);
+
+            int metadataVersion = metaMem.getInt(TableUtils.META_OFFSET_METADATA_VERSION);
+
+            // make sure we aren't duplicating work
+            CairoTable potentiallyExistingTable = tables.get(token.getDirName());
+            if (potentiallyExistingTable != null && potentiallyExistingTable.getMetadataVersion() > metadataVersion) {
+                LOG.debug().$("table in cache with newer version [table=")
+                        .$(token.getTableName()).$(", version=").$(potentiallyExistingTable.getMetadataVersion()).I$();
+                return;
+            }
+
+            // get basic metadata
+            int columnCount = metaMem.getInt(TableUtils.META_OFFSET_COUNT);
+
+            LOG.debug().$("reading columns [table=").$(token.getTableName())
+                    .$(", count=").$(columnCount)
+                    .I$();
+
+            table.setMetadataVersion(metadataVersion);
+
+            LOG.debug().$("set metadata version [table=").$(token.getTableName())
+                    .$(", version=").$(metadataVersion)
+                    .I$();
+
+            table.setPartitionBy(metaMem.getInt(TableUtils.META_OFFSET_PARTITION_BY));
+            table.setMaxUncommittedRows(metaMem.getInt(TableUtils.META_OFFSET_MAX_UNCOMMITTED_ROWS));
+            table.setO3MaxLag(metaMem.getLong(TableUtils.META_OFFSET_O3_MAX_LAG));
+            table.setTimestampIndex(metaMem.getInt(TableUtils.META_OFFSET_TIMESTAMP_INDEX));
+            table.setIsSoftLink(isSoftLink);
+
+            TableUtils.buildWriterOrderMap(metaMem, table.columnOrderMap, metaMem, columnCount);
+
+            // populate columns
+            for (int i = 0, n = table.columnOrderMap.size(); i < n; i += 3) {
+
+                int writerIndex = table.columnOrderMap.get(i);
+                if (writerIndex < 0) {
+                    continue;
+                }
+                int stableIndex = i / 3;
```

Comment:
nit: this requires some commentary. Perhaps elaborate on the structure of columnOrderMap?